File: apps/social/src/lib/coding-harness-probe.ts
/**
 * Coding harness probe — EnvoyDev-aligned Ready / Not ready.
 * Mirrors EnvoyGo `coding_harness_probe.dart`.
 */
import {
  codingHarnessToExtAgentId,
  type CodingHarnessId,
} from "@envoymesh/api";

export type HarnessProbeBadge = "ready" | "not-ready" | "checking";

export type HarnessProbeResult = {
  badge: HarnessProbeBadge;
  line?: string;
  hint?: string;
  installCommand?: string;
  installLink?: string;
};

type ProbeClient = {
  getEnvoyHarnessStatus?: () => Promise<{ state?: string; error?: string }>;
  getPiStatus?: () => Promise<{ state?: string; error?: string }>;
  probeExtAgent?: (params: {
    agentId: string;
  }) => Promise<{
    installState?: string;
    reachable?: boolean;
    installGuide?: {
      installed?: boolean;
      installCommand?: string;
      homepageUrl?: string;
      installLink?: string;
      startHint?: string;
      command?: string;
      commonIssues?: string[];
    };
  }>;
};

/** Collapse legacy install/unknown badges to Not ready for display. */
export function harnessProbeLabelKey(
  badge: HarnessProbeBadge | "install" | "unknown" | undefined,
): "ready" | "not-ready" | "checking" | undefined {
  if (!badge) return undefined;
  if (badge === "ready") return "ready";
  if (badge === "checking") return "checking";
  return "not-ready";
}

export async function probeCodingHarness(
  client: ProbeClient,
  harness: CodingHarnessId,
): Promise<HarnessProbeResult> {
  try {
    if (harness === "envoy-harness") {
      const get = client.getEnvoyHarnessStatus;
      if (!get) return { badge: "not-ready", line: "Status unavailable." };
      const s = await get();
      if (s.state === "ready") {
        return { badge: "ready", line: "Ready on this home computer." };
      }
      return {
        badge: "not-ready",
        line: s.error?.trim() || "Envoy Harness is not ready yet.",
        hint: s.error,
      };
    }

    if (harness === "pi") {
      const get = client.getPiStatus;
      if (!get) return { badge: "not-ready", line: "Status unavailable." };
      const s = await get();
      const state = s.state ?? "";
      if (state === "ready" || state === "stopped" || state === "starting") {
        return {
          badge: "ready",
          line:
            state === "ready"
              ? "Ready on this home computer."
              : "Will start when you send a message.",
        };
      }
      if (state === "disabled") {
        return {
          badge: "not-ready",
          line: "Disabled — enable in Settings → AI.",
          hint: "Pi is disabled. Enable it in Settings → AI.",
        };
      }
      if (state === "not-installed") {
        return {
          badge: "not-ready",
          line: "Not bundled on this home node (slim build).",
        };
      }
      return {
        badge: "not-ready",
        line: s.error?.trim() || "Pi is not ready.",
        hint: s.error,
      };
    }

    const agentId = codingHarnessToExtAgentId(harness);
    const probe = client.probeExtAgent;
    if (!agentId || !probe) {
      return { badge: "not-ready", line: "Could not confirm install status." };
    }
    const r = await probe({ agentId });
    const guide = r.installGuide;
    const cmd = guide?.installCommand?.trim();
    const link = guide?.installLink?.trim() || guide?.homepageUrl?.trim();
    const startHint = guide?.startHint?.trim();

    if (r.installState === "unsupported") {
      return {
        badge: "not-ready",
        line: "Cannot drive this agent on this machine yet.",
        hint: startHint,
      };
    }
    if (r.installState === "not-installed" || guide?.installed === false) {
      const issues = Array.isArray(guide?.commonIssues)
        ? guide.commonIssues.map((s) => String(s).trim()).filter(Boolean)
        : [];
      const hint =
        startHint ||
        issues[0] ||
        "Install this agent on your home computer.";
      return {
        badge: "not-ready",
        line: cmd || "Not installed",
        hint,
        installCommand: cmd,
        installLink: link,
      };
    }
    if (r.installState === "installed" || r.reachable) {
      if (r.reachable !== true && startHint) {
        return {
          badge: "not-ready",
          line: "Installed — needs a short setup step.",
          hint: startHint,
          installCommand: cmd,
          installLink: link,
        };
      }
      return {
        badge: "ready",
        line: guide?.command
          ? `Ready (${guide.command}).`
          : "Ready on this home computer.",
        hint: startHint,
      };
    }
    return {
      badge: "not-ready",
      line: startHint || "Could not confirm install status.",
      hint: startHint,
      installCommand: cmd,
      installLink: link,
    };
  } catch (e) {
    return {
      badge: "not-ready",
      line: e instanceof Error ? e.message : String(e),
    };
  }
}

/** True when the picker should open the install/resolve dialog. */
export function harnessNeedsResolve(
  badge: HarnessProbeBadge | "install" | "unknown" | undefined,
): boolean {
  return (
    badge === "not-ready" || badge === "install" || badge === "unknown"
  );
}

File: apps/envoygo/lib/coding/coding_harness_probe.dart
/// Probe Coding harness readiness — EnvoyDev-aligned Ready / Not ready.
library;

import '../services/product/node_service_client.dart';

/// Transient UI state while measuring; never a lasting verdict.
enum CodingHarnessProbeBadge {
  checking,
  ready,
  /// User-facing collapse of install / not-ready / unknown → show "Not ready".
  notReady,
}

/// Why a row is not ready (evidence behind the chip).
enum CodingHarnessNotReadyReason {
  /// Binary missing — run install steps.
  absent,
  /// Present but needs adapter / auth / warm-up (connector-style).
  connector,
  /// Missing env / API key.
  env,
  /// Product cannot drive this agent yet.
  ourGap,
  /// Probe failed or unknown.
  unlooked,
}

enum CodingHarnessGuideKind {
  steps,
  environment,
  app,
  nothing,
}

class CodingHarnessProbeResult {
  const CodingHarnessProbeResult({
    required this.badge,
    this.reason,
    this.line,
    this.hint,
    this.installCommand,
    this.installLink,
    this.startHint,
    this.guideKind = CodingHarnessGuideKind.steps,
  });

  final CodingHarnessProbeBadge badge;
  final CodingHarnessNotReadyReason? reason;
  /// One actionable line under the name (EnvoyDev row line).
  final String? line;
  final String? hint;
  final String? installCommand;
  final String? installLink;
  final String? startHint;
  final CodingHarnessGuideKind guideKind;

  bool get isReady => badge == CodingHarnessProbeBadge.ready;
  bool get needsResolve => badge == CodingHarnessProbeBadge.notReady;
}

/// Collapse wire evidence to the two user-facing verdicts (plus Checking…).
CodingHarnessProbeBadge codingHarnessUserBadge(CodingHarnessProbeBadge badge) {
  return switch (badge) {
    CodingHarnessProbeBadge.checking => CodingHarnessProbeBadge.checking,
    CodingHarnessProbeBadge.ready => CodingHarnessProbeBadge.ready,
    CodingHarnessProbeBadge.notReady => CodingHarnessProbeBadge.notReady,
  };
}

/// [wireId] is `envoy-harness` | `pi` | Tier B id (`claudecode`, `minimax-code`, …).
/// [extAgentId] overrides the Ext Agent probe id when it differs from [wireId].
Future<CodingHarnessProbeResult> probeCodingHarnessByWireId(
  NodeServiceClient client, {
  required String wireId,
  String? extAgentId,
}) async {
  final id = wireId.trim();
  try {
    if (id == 'envoy-harness') {
      final s = await client.getEnvoyHarnessStatus();
      final state = s['state']?.toString() ?? '';
      if (state == 'ready') {
        return const CodingHarnessProbeResult(
          badge: CodingHarnessProbeBadge.ready,
          line: 'Ready on this home computer.',
        );
      }
      if (state == 'disabled') {
        return CodingHarnessProbeResult(
          badge: CodingHarnessProbeBadge.notReady,
          reason: CodingHarnessNotReadyReason.env,
          guideKind: CodingHarnessGuideKind.app,
          line: 'Disabled — enable in Settings → AI.',
          hint: s['error']?.toString() ??
              'Configure Envoy Harness in Settings → AI.',
          startHint: 'Settings → AI',
        );
      }
      final err = s['error']?.toString().trim();
      return CodingHarnessProbeResult(
        badge: CodingHarnessProbeBadge.notReady,
        reason: CodingHarnessNotReadyReason.unlooked,
        guideKind: CodingHarnessGuideKind.app,
        line: (err != null && err.isNotEmpty)
            ? err
            : 'Envoy Harness is not ready yet.',
        hint: (err != null && err.isNotEmpty)
            ? err
            : 'Envoy Harness is not ready yet.',
        startHint: 'Settings → AI',
      );
    }

    if (id == 'pi') {
      final s = await client.getPiStatus();
      final state = s['state']?.toString() ?? '';
      final err = s['error']?.toString().trim();
      // stopped/starting: sendToPi / restartPi can bring it up → Ready for picker.
      if (state == 'ready' || state == 'stopped' || state == 'starting') {
        return CodingHarnessProbeResult(
          badge: CodingHarnessProbeBadge.ready,
          line: state == 'ready'
              ? 'Ready on this home computer.'
              : 'Will start when you send a message.',
        );
      }
      if (state == 'disabled') {
        return const CodingHarnessProbeResult(
          badge: CodingHarnessProbeBadge.notReady,
          reason: CodingHarnessNotReadyReason.env,
          guideKind: CodingHarnessGuideKind.app,
          line: 'Disabled — enable in Settings → AI.',
          hint: 'Pi is disabled. Enable it in Settings → AI (Coding agents).',
          startHint: 'Settings → AI → Coding agents',
        );
      }
      if (state == 'not-installed') {
        return const CodingHarnessProbeResult(
          badge: CodingHarnessProbeBadge.notReady,
          reason: CodingHarnessNotReadyReason.absent,
          guideKind: CodingHarnessGuideKind.nothing,
          line: 'Not bundled on this home node (slim build).',
          hint: 'Pi sidecar is not bundled on this home node (slim build).',
          startHint: 'Use a full desktop build, or pick Envoy / another agent.',
        );
      }
      return CodingHarnessProbeResult(
        badge: CodingHarnessProbeBadge.notReady,
        reason: CodingHarnessNotReadyReason.unlooked,
        guideKind: CodingHarnessGuideKind.app,
        line: (err != null && err.isNotEmpty)
            ? err
            : 'Pi is not ready.',
        hint: (err != null && err.isNotEmpty)
            ? err
            : 'Pi is not ready. Check Settings → AI model config.',
        startHint: 'Settings → AI',
      );
    }

    final agentId = (extAgentId ?? id).trim();
    final r = await client.probeExtAgent(agentId: agentId);
    final installState = r['installState']?.toString() ?? '';
    final guide = r['installGuide'];
    String? cmd;
    String? link;
    String? startHint;
    String? binaryName;
    if (guide is Map) {
      cmd = guide['installCommand']?.toString().trim();
      link = guide['installLink']?.toString().trim() ??
          guide['homepageUrl']?.toString().trim();
      startHint = guide['startHint']?.toString().trim();
      binaryName = guide['command']?.toString().trim();
    }

    if (installState == 'unsupported') {
      return CodingHarnessProbeResult(
        badge: CodingHarnessProbeBadge.notReady,
        reason: CodingHarnessNotReadyReason.ourGap,
        guideKind: CodingHarnessGuideKind.nothing,
        line: 'EnvoyGo cannot drive this agent on this machine yet.',
        hint: startHint ?? 'This agent is not supported on this platform.',
        startHint: startHint,
      );
    }

    if (installState == 'not-installed' ||
        (guide is Map && guide['installed'] == false)) {
      final issues = <String>[];
      if (guide is Map && guide['commonIssues'] is List) {
        for (final row in guide['commonIssues'] as List) {
          final s = row?.toString().trim() ?? '';
          if (s.isNotEmpty) issues.add(s);
        }
      }
      final lineCmd =
          (cmd != null && cmd.isNotEmpty) ? cmd : 'Not installed';
      final hint = (startHint != null && startHint.isNotEmpty)
          ? startHint
          : (issues.isNotEmpty
              ? issues.first
              : 'Install this agent on your home computer.');
      return CodingHarnessProbeResult(
        badge: CodingHarnessProbeBadge.notReady,
        reason: CodingHarnessNotReadyReason.absent,
        guideKind: CodingHarnessGuideKind.steps,
        line: lineCmd,
        hint: hint,
        installCommand: (cmd != null && cmd.isNotEmpty) ? cmd : null,
        installLink: (link != null && link.isNotEmpty) ? link : null,
        startHint: startHint ?? hint,
      );
    }

    if (installState == 'installed' || r['reachable'] == true) {
      // Installed but not reachable → connector / auth style guidance.
      if (r['reachable'] != true &&
          startHint != null &&
          startHint.isNotEmpty) {
        final looksEnv = startHint.contains('API_KEY') ||
            startHint.contains('auth login') ||
            startHint.contains('environment');
        return CodingHarnessProbeResult(
          badge: CodingHarnessProbeBadge.notReady,
          reason: looksEnv
              ? CodingHarnessNotReadyReason.env
              : CodingHarnessNotReadyReason.connector,
          guideKind: looksEnv
              ? CodingHarnessGuideKind.environment
              : CodingHarnessGuideKind.steps,
          line: looksEnv
              ? startHint
              : 'Installed — needs a short setup step.',
          hint: startHint,
          installCommand: (cmd != null && cmd.isNotEmpty) ? cmd : null,
          installLink: (link != null && link.isNotEmpty) ? link : null,
          startHint: startHint,
        );
      }
      return CodingHarnessProbeResult(
        badge: CodingHarnessProbeBadge.ready,
        line: binaryName != null && binaryName.isNotEmpty
            ? 'Ready ($binaryName).'
            : 'Ready on this home computer.',
        hint: startHint,
      );
    }

    return CodingHarnessProbeResult(
      badge: CodingHarnessProbeBadge.notReady,
      reason: CodingHarnessNotReadyReason.unlooked,
      guideKind: CodingHarnessGuideKind.steps,
      line: startHint ?? 'Could not confirm install status.',
      hint: startHint ?? 'Could not confirm install status.',
      installCommand: (cmd != null && cmd.isNotEmpty) ? cmd : null,
      installLink: (link != null && link.isNotEmpty) ? link : null,
      startHint: startHint,
    );
  } catch (e) {
    return CodingHarnessProbeResult(
      badge: CodingHarnessProbeBadge.notReady,
      reason: CodingHarnessNotReadyReason.unlooked,
      guideKind: CodingHarnessGuideKind.app,
      line: e.toString().replaceFirst('Exception: ', ''),
      hint: e.toString().replaceFirst('Exception: ', ''),
    );
  }
}
